cache/cache_reservation_manager.h

```diff
@@ -132,7 +154,41 @@ class CacheReservationManager
   template <CacheEntryRole R>
   Status MakeCacheReservation(
       std::size_t incremental_memory_used,
-      std::unique_ptr<CacheReservationHandle<R>> *handle);
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle<R>>
+          *handle);
+
+  // One of the two ways of releasing cache reservation, used with
+  // MakeCacheReservation(). It's usually called on the destruction of a
+  // CacheReservationManager::CacheReservationHandle<R> handle, which achieves
+  // better RAII per cache reservation.
+  // See CacheReservationManager::UpdateCacheReservation()
+  // for the other way.
+  //
+  // Use ONLY one of them to prevent unexpected behavior.
+  //
+  // Release dummy entries in the cache for the incremental memory usage
+  // to match the size of total dummy entries with the least multiple of
+  // kSizeDummyEntry greater than or equal to the total memory used.
+  //
+  // WARNING: The passed-in incremental_memory_used should be resulted from
+  //          calling GetIncrementalMemoryUsed() on some
+  //          CacheReservationManager::CacheReservationHandle<R>> object
+  //          created by MakeCacheReservation() and hasn't had its cache
+  //          reservation released
+  //
+  // @param incremental_memory_used The number of bytes previously increased in
+  //        memory usage.
+  //
+  //        Calling GetTotalMemoryUsed() afterward will return the total memory
+  //        decreased by this number.
+  //
+  //        Since the class is NOT thread-safe, external synchronization in
+  //        calling ReleaseCacheReservation() is needed if you want
+  //        GetTotalMemoryUsed() indeed returns the latest memory used.
+  //
+  // @return It always returns Status::OK()
+  template <CacheEntryRole R>
+  Status ReleaseCacheReservation(std::size_t incremental_memory_used);
```

Comment:
Have you tried making this private (or protected if you follow another suggestion -- see below)? the inner class might be able to access a private function of the outer class even without declaring it a friend.

Reply:
> the inner class might be able to access a private function of the outer class even without declaring it a friend.

Yeah - the issue I ran into is, unlike Java, C++ needs to store a raw pointer or reference of the enclosing class in child class https://stackoverflow.com/questions/486099/can-inner-classes-access-private-variables.

I tried to store smart pointer (std::shared_ptr) instead and couldn't get it work ...

Reply:
The error I got when making it private is in the access from CacheReservationManagerThreadSafeWrapper::ReleaseCacheReservation, which isn't an inner class. After working around that it seemed to work.

Reply:
Done by making ReleaseCacheReservation protected as suggested.